### PR TITLE
[Rsdk-572] fix readings for all movement sensor, return errors for unimplemented methods

### PR DESCRIPTION
**Pull request by randhid**
Now returning unimplemented methods for movement sensors that don't implement those methods.  Doesn't hang usb on close

Warning added from nmea parser (what causes a nil gps location) that will look like: 
Screenshot 2022-09-30 at 17 21 11

### CODE REVIEW

File: components/movementsensor/utils.go [hunk 1]
```diff
@@ -47,3 +49,37 @@ func GetHeading(gps1, gps2 *geo.Point, yawOffset float64) (float64, float64, flo
...
+	case compStr:
+		return errors.New("CompassHeading Unimplemented")
+	case accuracyStr:
+		return errors.New("Accuracy Unimplemented")
+	case readStr:
+		return errors.New("Readings Unimplemented")
+	default:
+		return errs.Errorf("unknown method name %s used for creating unimerror", method)
+	}
+}
```
Comment (maximpertsov) [inline components/movementsensor/utils.go:85]: I think it's better to just define concrete error for each kind of unimplemented method. This gives us compile-time checks against misspells.

e.g. 

```
var ErrMethodUnimplementedPosition = errors.New("Position Unimplemented")
var ErrMethodUnimplementedOrientation = return errors.New("Orientation Unimplemented")

// etc
```

**Review by maximpertsov — CHANGES_REQUESTED** [review]

File: components/movementsensor/gpsnmea/gpsnmea.go [hunk 2]
```diff
@@ -48,7 +48,7 @@ type I2CAttrConfig struct {
 
 // Validate ensures all parts of the config are valid.
 func (cfg *AttrConfig) Validate(path string) error {
-	if cfg.Board == "" {
+	if cfg.Board == "" && cfg.ConnectionType == i2cStr {
```
[the remark this answers is not in the document]
Comment (randhid) [reply components/movementsensor/gpsnmea/gpsnmea.go:51]: yeah I was testing on my laptop for the wit imu/gps

File: components/movementsensor/gpsnmea/serial.go [hunk 3]
```diff
@@ -152,54 +152,57 @@ func (g *SerialNMEAMovementSensor) GetCorrectionInfo() (string, uint) {
...
-	return 0, g.lastError
+	return 0, movementsensor.ErrMethodUnimplemented("CompassHeading")
 }
 
 // ReadFix returns Fix quality of MovementSensor measurements.
 func (g *SerialNMEAMovementSensor) ReadFix(ctx context.Context) (int, error) {
 	g.mu.RLock()
 	defer g.mu.RUnlock()
-	return g.data.fixQuality, g.lastError
+	return g.data.fixQuality, nil
```
[the remark this answers is not in the document]
Comment (maximpertsov) [reply components/movementsensor/gpsnmea/serial.go:198]: discussed IRL - last captured error shouldn't really be a factor if read fix is returned successfully

File: components/movementsensor/gpsnmea/serial.go [hunk 4]
```diff
@@ -152,54 +152,57 @@ func (g *SerialNMEAMovementSensor) GetCorrectionInfo() (string, uint) {
...
+	}
 	return g.data.location, g.data.alt, g.lastError
 }
 
 // Accuracy returns the accuracy, hDOP and vDOP.
 func (g *SerialNMEAMovementSensor) Accuracy(ctx context.Context) (map[string]float32, error) {
 	g.mu.RLock()
 	defer g.mu.RUnlock()
-	return map[string]float32{"hDOP": float32(g.data.hDOP), "vDOP": float32(g.data.vDOP)}, g.lastError
+	return map[string]float32{"hDOP": float32(g.data.hDOP), "vDOP": float32(g.data.vDOP)}, nil
```
[the remark this answers is not in the document]
Comment (maximpertsov) [reply components/movementsensor/gpsnmea/serial.go:165]: makes sense, we can leave nil - sounds similar to https://github.com/viamrobotics/rdk/pull/1420#discussion_r986116796

File: components/movementsensor/gpsnmea/nmeaParser.go [hunk 5]
```diff
@@ -25,6 +27,10 @@ type gpsData struct {
 	fixQuality int
 }
 
+func errInvalidFix(sentenceType, badFix, goodFix string) error {
+	return errors.Errorf("\n type %s sentence fix is not valid have: %s  want %s", sentenceType, badFix, goodFix)
+}
```
[the remark this answers is not in the document]
Comment (randhid) [reply components/movementsensor/gpsnmea/nmeaParser.go]: missed this but updated code on my side.

File: components/movementsensor/movementsensor.go [hunk 6]
```diff
@@ -130,32 +130,32 @@ func Readings(ctx context.Context, g MovementSensor) (map[string]interface{}, er
 	readings := map[string]interface{}{}
 
 	pos, altitide, err := g.Position(ctx)
-	if err != nil {
+	if err != nil && err.Error() != ErrMethodUnimplementedPosition.Error() {
```
Comment (maximpertsov) [inline components/movementsensor/movementsensor.go:133]: nit, i think it's more idiomatic to do this

```suggestion
	if err != nil && errors.Is(err, ErrMethodUnimplementedPosition) {
```

File: components/movementsensor/movementsensor.go [hunk 7]
```diff
@@ -130,32 +130,32 @@ func Readings(ctx context.Context, g MovementSensor) (map[string]interface{}, er
...
 		return nil, err
 	}
 	readings["position"] = pos
 	readings["altitide"] = altitide
 
 	vel, err := g.LinearVelocity(ctx)
-	if err != nil {
+	if err != nil && err.Error() != ErrMethodUnimplementedLinearVelocity.Error() {
```
Comment (maximpertsov) [inline components/movementsensor/movementsensor.go:140]:
```suggestion
	if err != nil && errors.Is(err, ErrMethodUnimplementedLinearVelocity) {
```

File: components/movementsensor/movementsensor.go [hunk 8]
```diff
@@ -130,32 +130,32 @@ func Readings(ctx context.Context, g MovementSensor) (map[string]interface{}, er
...
 		return nil, err
 	}
 	readings["linear_velocity"] = vel
 
 	avel, err := g.AngularVelocity(ctx)
-	if err != nil {
+	if err != nil && err.Error() != ErrMethodUnimplementedAngularVelocity.Error() {
```
Comment (maximpertsov) [inline components/movementsensor/movementsensor.go:146]:
```suggestion
	if err != nil && errorsIs(err, ErrMethodUnimplementedAngularVelocity) {
```

File: components/movementsensor/movementsensor.go [hunk 9]
```diff
@@ -130,32 +130,32 @@ func Readings(ctx context.Context, g MovementSensor) (map[string]interface{}, er
...
 		return nil, err
 	}
 	readings["angular_velocity"] = avel
 
 	compass, err := g.CompassHeading(ctx)
-	if err != nil {
+	if err != nil && err.Error() != ErrMethodUnimplementedCompassHeading.Error() {
```
Comment (maximpertsov) [inline components/movementsensor/movementsensor.go:152]:
```suggestion
	if err != nil && errors.Is(err, ErrMethodUnimplementedCompassHeading) {
```

File: components/movementsensor/movementsensor.go [hunk 10]
```diff
@@ -130,32 +130,32 @@ func Readings(ctx context.Context, g MovementSensor) (map[string]interface{}, er
...
 		return nil, err
 	}
 	readings["compass"] = compass
 
 	ori, err := g.Orientation(ctx)
-	if err != nil {
+	if err != nil && err.Error() != ErrMethodUnimplementedOrientation.Error() {
```
Comment (maximpertsov) [inline components/movementsensor/movementsensor.go:158]:
```suggestion
	if err != nil && errors.Is(err, ErrMethodUnimplementedOrientation) {
```

**Review by maximpertsov — APPROVED** [review]

**Pull request merged**